nupm/install.nu

```diff
@@ -0,0 +1,56 @@
+use std log
+
+def throw-error [
+    error: string
+    text?: string
+    --span: record<start: int, end: int>
+] {
+    let error = $"(ansi red_bold)($error)(ansi reset)"
+
+    if $span == null {
+        error make --unspanned { msg: $error }
+    }
+
+    error make {
+        msg: $error
+        label: {
+            text: ($text | default "this caused an internal error")
+            start: $span.start
+            end: $span.end
+        }
+    }
+}
+
+def nupm-home [] {
+    $env.NUPM_HOME? | default (
+        $env.XDG_DATA_HOME?
+        | default ($nu.home-path | path join ".local" "share")
+        | path join "nupm"
+    )
+}
+
+# install a Nushell package
+export def main [
+    --path: path  # the path to the local source of the package (defaults to the current directory)
+] {
+    let path = ($path | default . | path expand)
+
+    # NOTE: here, we suppose that the package file exists and is valid
+    let package = ($path | path join "package.nuon" | open)
+
+    let destination = (nupm-home | path join $package.name)
+    if not ($destination | path exists) {
+        log info $"installing package ($package.name)"
+        log debug $"source: ($path)"
+        log debug $"destination: ($destination)"
+        cp --recursive $path $destination
+
+        if ($destination | path join ".git" | path exists) {
+            log debug $"removing .git directory from ($destination)"
+            rm --recursive ($destination | path join ".git")
+        }
+    } else {
+        # TODO: add support for updating / reinstalling a package in that case
+        throw-error $"package ($package.name) is already installed"
```

Comment:
Instead of throwing an error maybe we could just log that package is already installed to make the operation idempotent

Reply:
should be good in 3ff7e91 :relieved: 

i'm curious to know what *idempotent* means here :yum:

Reply:
Run the same command multiple times, get same result every time

Reply:
oooh i see :ok_hand: 

i'm not sure how it applies here because
- on the first `nupm install`, the package would be installed
- on the next ones, `nupm install` would give a warning

is this *idempotent*?
and if it's not, how can we make it so? :open_mouth:

Reply:
I don't think we even need a warning here. If we define the installation command as 'I want these packages installed on my system' and not 'I want to perform installation operation on these packages' (so expression of desired state and not operations to perform) then it's only logical that installing packages that are already present on the system is noop

Reply:
oooooh i see what you mean now, thanks :+1: